### ISSUE TITLE
Xlint:path error with JUNG, Gradle and Error Prone

### ISSUE THREAD

**Issue by jbduncan** [google/error-prone#466]
Hi Error Prone team.

This is an issue which I posted already on https://github.com/jrtom/jung/issues/66. @jrtom seems to be at a loss as to what's causing it, so I wonder if someone on the Error Prone team can help me.

I've pasted the original issue text below so that people don't have to follow the link to read it.

> When I try to run `gradle run` with the following `build.gradle`
> 
> ``` gradle
> plugins {
>   id 'application'
>   id 'java'
>   id 'net.ltgt.errorprone' version '0.0.8'
> }
> 
> group 'org.jbduncan'
> version '1.0-SNAPSHOT'
> 
> sourceCompatibility = 1.8
> 
> mainClassName = 'org.jbduncan.helloworld.HelloWorld'
> 
> repositories {
>   jcenter()
> }
> 
> dependencies {
>   compile 'net.sf.jung:jung-graph-impl:2.1.1'
>   errorprone 'com.google.errorprone:error_prone_core:2.0.11'
> }
> 
> tasks.withType(JavaCompile) {
>   options.encoding = 'UTF-8'
>   options.compilerArgs = [
>     '-Xlint:all',
>     '-Werror'
>   ]
> }
> ```
> 
> and this class
> 
> ``` java
> package org.jbduncan.helloworld;
> 
> public final class HelloWorld {
>   public static void main(String[] args) {
>     System.out.println("Hello, world!");
>   }
> }
> ```
> 
> it emits the following message.
> 
> ```
> :compileJava                                                                     
> warning: [path] bad path element "C:\Users\Jonathan\.gradle\caches\modules-2\files-2.1\net.sf.jung\jung-graph-impl\2.1.1\8293acb2ab4c00a3939cb99a8751e5d38a4299dc\jung-api-2.1.1.jar": no such file or directory
> warning: [path] bad path element "C:\Users\Jonathan\.gradle\caches\modules-2\files-2.1\net.sf.jung\jung-graph-impl\2.1.1\8293acb2ab4c00a3939cb99a8751e5d38a4299dc\guava-19.0.jar": no such file or directory
> warning: [path] bad path element "C:\Users\Jonathan\.gradle\caches\modules-2\files-2.1\net.sf.jung\jung-api\2.1.1\e47ee4efdfacce12f0af620747d9d0e44bf2eaa4\guava-19.0.jar": no such file or directory
> error: warnings found and -Werror specified
> 1 error                     
> 3 warnings
> :compileJava FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':compileJava'.
> > Compilation failed with exit code 1; see the compiler error output for details.
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
> 
> BUILD FAILED
> 
> Total time: 5.488 secs
> ```
> 
> When I investigate the contents of my Gradle cache, I can see that, indeed, the jar files mentioned above simply don't exist.
> 
> When I look into `C:\Users\Jonathan\.gradle\caches\modules-2\files-2.1\net.sf.jung\jung-graph-impl\2.1.1\8293acb2ab4c00a3939cb99a8751e5d38a4299dc\jung-graph-impl-2.1.1.jar`, the `META-INF/MANIFEST.MF` contains this text.
> 
> ```
> Manifest-Version: 1.0
> Archiver-Version: Plexus Archiver
> Created-By: Apache Maven 3.0.5
> Built-By: jrtom
> Build-Jdk: 1.8.0-google-v7
> Class-Path: jung-api-2.1.1.jar guava-19.0.jar
> ```
> 
> I get the same problem with JUNG versions 2.1 and 2.0, and no other library with dependencies I've tried seems to have this problem. 
> 
> However the problem goes away when I exclude from `build.gradle` the lines `id 'net.ltgt.errorprone' version '0.0.8'` and `errorprone 'com.google.errorprone:error_prone_core:2.0.11'`, so it's not clear to me if this is a problem with JUNG or if it's [gradle-errorprone-plugin](https://github.com/tbroyer/gradle-errorprone-plugin) or Error Prone related.

(Since writing the original issue, I upgraded from Error Prone 2.0.11 to 2.0.13, but that didn't seem to solve my problem.)

**Comment by tbroyer** [comment]
Apparently, it's the expected behavior of JavaC; see: http://bugs.java.com/bugdatabase/view_bug.do?bug_id=6485027 (and related, specifically http://bugs.java.com/bugdatabase/view_bug.do?bug_id=4212732), along with https://github.com/google/error-prone-javac/blob/90149b22e202e7b24bfafc80df7509c8fb490ee7/src/jdk.compiler/share/classes/com/sun/tools/javac/file/Locations.java#L317-L329 (which is the post-8 pre-9 version of JavaC used by ErrorProne), https://github.com/google/error-prone-javac/blob/90149b22e202e7b24bfafc80df7509c8fb490ee7/test/tools/javac/Paths/Class-Path.sh, and https://github.com/google/error-prone-javac/blob/90149b22e202e7b24bfafc80df7509c8fb490ee7/test/tools/javac/Paths/TestCompileJARInClassPath.java (among others).

So the question actually is: why isn't the error happening without ErrorProne?

**Comment by cushon** [comment]
> So the question actually is: why isn't the error happening without ErrorProne?

Error Prone invokes javac using internal APIs, and gradle uses JSR 199 which doesn't handle `-Xlint:path` correctly: http://mail.openjdk.java.net/pipermail/compiler-dev/2016-October/010455.html

**Comment by tbroyer** [comment]
FWIW, running Gradle with `--debug`, I extracted the arguments it passes to JavaC:
```
17:33:04.984 [DEBUG] [org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler] Compiler arguments: -source 1.8 -target 1.8 -d /…/build/classes/main -g -encoding UTF-8 -sourcepath /…/build/tmp/compileJava/emptySourcePathRef -Xlint:all -Werror -classpath /…/.gradle/caches/modules-2/files-2.1/net.sf.jung/jung-graph-impl/2.1.1/8293acb2ab4c00a3939cb99a8751e5d38a4299dc/jung-graph-impl-2.1.1.jar:/…/.gradle/caches/modules-2/files-2.1/net.sf.jung/jung-api/2.1.1/e47ee4efdfacce12f0af620747d9d0e44bf2eaa4/jung-api-2.1.1.jar:/…/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/19.0/6ce200f6b23222af3d8abb6b6459e6c44f4bb0e9/guava-19.0.jar /…/src/main/java/org/jbduncan/helloworld/HelloWorld.java -XDuseUnsharedTable=true
```
Now running JavaC with those same arguments gives the same error as with ErrorProne:
```
$ javac -source 1.8 -target 1.8 -d /…/build/classes/main -g -encoding UTF-8 -sourcepath /…/build/tmp/compileJava/emptySourcePathRef -Xlint:all -Werror -classpath /…/.gradle/caches/modules-2/files-2.1/net.sf.jung/jung-graph-impl/2.1.1/8293acb2ab4c00a3939cb99a8751e5d38a4299dc/jung-graph-impl-2.1.1.jar:/…/.gradle/caches/modules-2/files-2.1/net.sf.jung/jung-api/2.1.1/e47ee4efdfacce12f0af620747d9d0e44bf2eaa4/jung-api-2.1.1.jar:/…/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/19.0/6ce200f6b23222af3d8abb6b6459e6c44f4bb0e9/guava-19.0.jar /…/src/main/java/org/jbduncan/helloworld/HelloWorld.java -XDuseUnsharedTable=true
warning: [path] bad path element "/…/.gradle/caches/modules-2/files-2.1/net.sf.jung/jung-graph-impl/2.1.1/8293acb2ab4c00a3939cb99a8751e5d38a4299dc/jung-api-2.1.1.jar": no such file or directory
warning: [path] bad path element "/…/.gradle/caches/modules-2/files-2.1/net.sf.jung/jung-graph-impl/2.1.1/8293acb2ab4c00a3939cb99a8751e5d38a4299dc/guava-19.0.jar": no such file or directory
warning: [path] bad path element "/…/.gradle/caches/modules-2/files-2.1/net.sf.jung/jung-api/2.1.1/e47ee4efdfacce12f0af620747d9d0e44bf2eaa4/guava-19.0.jar": no such file or directory
error: warnings found and -Werror specified
1 error
3 warnings
```

Note that the arguments passed by Gradle are the same whether you use the gradle-errorprone-plugin or not.

So, from my point of view, it's a bug in Gradle (and other build tools; I could similarly reproduce this with Maven: build succeeds, running with `-X` to get the arguments and then running JavaC on the command-line with the same arguments fails, with the same output, modulo JAR paths; Ant correctly fails the compilation though), and it's a bug in JUNG for deploying such JARs to a Maven repository.

It's arguable too that Gradle and Maven explicitly ignore such errors by default *because* there are people deploying such JARs, and I could then have a look at doing the same in the gradle-errorprone-plugin. It's interesting to note that Maven has the same behavior though: compiles without complaining with standard maven-compiler-plugin, fails the build when using ErrorProne.

Also, unsurprisingly, compilation works with Bazel, as it extracts an _ijar_ and does not directly pass the JARs to JavaC/ErrorProne.

For the record, here's the Maven POM I used to replicate the issue with Maven:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">

  <!-- POM file generated with GWT webAppCreator -->
  <modelVersion>4.0.0</modelVersion>
  <groupId>test</groupId>
  <artifactId>Test</artifactId>
  <version>1.0-SNAPSHOT</version>
  <name>test.Test</name>

  <properties>
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
  </properties>

<dependencies>
<dependency>
<groupId>net.sf.jung</groupId>
<artifactId>jung-graph-impl</artifactId>
<version>2.1.1</version>
</dependency>
</dependencies>

  <build>
    <plugins>
<plugin>
<groupId>org.apache.maven.plugins</groupId>
<artifactId>maven-compiler-plugin</artifactId>
<version>3.5.1</version>
<configuration>
  <compilerArgs>
    <arg>-Xlint:all</arg>
    <arg>-Werror</arg>
  </compilerArgs>
  <showWarnings>true</showWarnings>
  <showDeprecations>true</showDeprecations>
<compilerId>javac-with-errorprone</compilerId>
        <forceJavacCompilerUse>true</forceJavacCompilerUse>
</configuration>
      <dependencies>
        <dependency>
          <groupId>org.codehaus.plexus</groupId>
          <artifactId>plexus-compiler-javac-errorprone</artifactId>
          <version>2.5</version>
        </dependency>
        <!-- override plexus-compiler-javac-errorprone's dependency on
             Error Prone with the latest version -->
        <dependency>
          <groupId>com.google.errorprone</groupId>
          <artifactId>error_prone_core</artifactId>
          <version>2.0.13</version>
        </dependency>
      </dependencies>
</plugin>
    </plugins>
  </build>
</project>
```

and the Ant `build.xml` (using the JARs downloaded by Maven):
```xml
<project basedir=".">
  <target name="javac">
    <javac srcdir="src/main/java" destdir="target/classes" source="1.8" target="1.8" encoding="UTF-8" includeantruntime="false">
      <compilerarg value="-Xlint:all"/>
      <compilerarg value="-Werror"/>
      <classpath>
        <pathelement location="/…/.m2/repository/net/sf/jung/jung-graph-impl/2.1.1/jung-graph-impl-2.1.1.jar" />
        <pathelement location="/…/.m2/repository/net/sf/jung/jung-api/2.1.1/jung-api-2.1.1.jar" />
        <pathelement location="/…/.m2/repository/com/google/guava/guava/19.0/guava-19.0.jar"/>
      </classpath>
    </javac>
  </target>
</project>
```

and finally the Bazel `WORKSPACE`:
```python
maven_jar(
	name = "net_sf_jung_jung_graph_impl",
	artifact = "net.sf.jung:jung-graph-impl:2.1.1",
)
maven_jar(
	name = "net_sf_jung_jung_api",
	artifact = "net.sf.jung:jung-api:2.1.1",
)
maven_jar(
	name = "com_google_guava_guava",
	artifact = "com.google.guava:guava:19.0",
)
```
and `BUILD`
```python
java_binary(
	name = 'helloworld',
	srcs = glob(['src/main/java/**/*.java']),
	deps = [
		'@net_sf_jung_jung_graph_impl//jar',
		'@net_sf_jung_jung_api//jar',
		'@com_google_guava_guava//jar',
	],
	main_class = 'org.jbduncan.helloworld.HelloWorld',
	javacopts = ['-Xlint:all', '-Werror'],
)
```

**Comment by tbroyer** [comment]
FWIW, using `<forceJavacCompilerUse>true</forceJavacCompilerUse>` without ErrorProne will break a Maven build (respects `-Xlint:path`, from the `-Xlint:all`).

And similarly for Gradle when using `options.fork = true` and `options.forkOptions.executbale = 'javac'`.

**Comment by cushon** [comment]
The fork/forceJavacCompilerUse options cause the javac CLI to be used instead of JSR 199, right? That would get the same non-buggy option handling Error Prone calls into.

> Also, unsurprisingly, compilation works with Bazel, as it extracts an ijar and does not directly pass the JARs to JavaC/ErrorProne.

ijar deletes MANIFESTs from the jars it processes, which I consider to be a feature.

**Comment by jbduncan** [comment]
Hmm, sounds to me like this is half a bug in JUNG and half a bug in JSR 199. Is my understanding correct?

What I've not yet understood is, if it's indeed half a bug in JUNG, how could I or @jrtom fix it on the JUNG side?

I see that @cushon has just raised the JSR 199 side of it with the OpenJDK devs at http://mail.openjdk.java.net/pipermail/compiler-dev/2016-October/010455.html. Does this mean we now wait until it's fixed on their end?

I also see that @tbroyer seems to think it's a bug with Gradle. Has this been raised on the Gradle forums yet? If not, I'd be happy to link this issue in the [topic](https://discuss.gradle.org/t/unable-to-use-xlint-all-in-a-gradle-project-with-gradle-errorprone-plugin-and-jung-library/20038?u=jbduncan) I raised.

**Comment by cushon** [comment]
I don't think there's a gradle bug here beyond the underlying JSR 199 bug.

I suggest having JUNG not set `Class-Path` in the jar's manifest.  Removing `<addClasspath>` from the pom should do it: https://github.com/jrtom/jung/blob/6c0a8e568a2215c15a8195f3ec75321f6241ce48/pom.xml#L152

Setting `Class-Path` makes assumptions about the filesystem layout of build tools that will be using the jars, and interacts badly with other dependency resolution mechanisms.

**Comment by tbroyer** [comment]
@cushon forceJavacCompilerUse forces the use of the com.sun.….Main class, and fork in Gradle will fork a cli javac process. So yes, they force not using jsr199, so bypass the bug.

> ijar deletes MANIFESTs from the jars it processes, which I consider to be a feature.

👍

@jbduncan Both Maven and Gradle use jsr199 by default, so indeed they're not direct at fault. FWIW, I believe old versions of Gradle allowed using JavaC Main the same way forceJavacCompilerUse does in Maven, this is no longer an option.

**Comment by jbduncan** [comment]
Thank you everyone, @cushon and @tbroyer included. It seems it was a bug in how the JUNG jar was being built, which [I've submitted a fix for](https://github.com/jrtom/jung/pull/75), so I'm closing this issue for now.

@cushon, on a related note, have you received any news regarding the bug in JSR 199 which you queried about at http://mail.openjdk.java.net/pipermail/compiler-dev/2016-October/010455.html?
